### ISSUE TITLE
Diagram metadata

### ISSUE THREAD

**Issue by vanillajonathan** [gaphor/gaphor#803]
I would like to set metadata.

- A short description of what the diagram represents.
- A URL to a website.
- Author name and email address.
- Revision, version, or last updated at.
- Created on date.
- Updated on date.
- License, maybe in the form of a SPDX short identifier such as [CC-PDDC](https://spdx.org/licenses/CC-PDDC.html).

**Comment by danyeaw** [comment]
UML/SysML species that you can attach a Comment element to a diagram itself.

![image](https://user-images.githubusercontent.com/10014976/118344637-38f15e00-b4fd-11eb-9e4e-de58c7c15784.png)

Gaphor currently doesn't support the diagram frame, or attaching comments to the diagram itself. If we could attach a comment to a diagram would this provide what you are looking for? Then you can define any information you want to show on each diagram.

**Comment by vanillajonathan** [comment]
> UML/SysML species that you can attach a Comment element to a diagram itself.

I wasn't are of that.

> If we could attach a comment to a diagram would this provide what you are looking for?

I don't know. Maybe. I just noticed there were no way to set any metadata and thought there ought to be a way to so. I had in mind a dialog with predefined fields but maybe this could work too.

**Comment by amolenaar** [comment]
@vanillajonathan What do you want do with the metadata?

**Comment by vanillajonathan** [comment]
Nothing in particular in this point of time, I just figured there ought to be some sort of metadata support.

So if someone comes across my file, this person can know what version it is, find a URL where he can find the latest version if there is any newer version, and if I made any mistakes it should be possible for someone to contact me over email to tell me.

**Comment by amolenaar** [comment]
@vanillajonathan I see. I would expect this file to be in a (git) repository or something, so you should be able to obtain all meta data from there as well.

**Comment by vanillajonathan** [comment]
If it gets printed on a paper then it might be useful to have the metadata printed on the paper.

**Comment by amolenaar** [comment]
We have an really nice Sphinx extension in Gaphor 2.7.

**Comment by vanillajonathan** [comment]
Sphinx is not what I want, because the image may be printed or it may be embedded on a website or redistributed. I want a metadata block in the corner of the image.

https://www.google.com/search?q=engineer+drawing+title+block&tbm=isch
https://en.wikipedia.org/wiki/Engineering_drawing#Title_block

**Comment by amolenaar** [comment]
@vanillajonathan Can you help us out with a mockup of what you want to see? I think some properties (company name, company logo) only have to be defined once, while diagram name, diagram version change per diagram. What about the author/creator name? Should it contain a date as well?

**Comment by vanillajonathan** [comment]
First off this needs to be be persisted to the `.gaphor` file as metadata.
Then that metadata can be read to be embedded as a block in the diagram.

Something like this:
![image](https://user-images.githubusercontent.com/10222521/144836946-29f81d59-3296-454f-ac25-65973ea998e3.png)

**Comment by amolenaar** [comment]
@vanillajonathan I was thinking about creating a `Metadata` class in the core model. It can contain the metadata information. It's a normal element and will show up in the tree view. It can be dragged in any diagram.

**Comment by vanillajonathan** [comment]
The application menu could have an entry called "Edit Metadata" which would bring up a Metadata Editor dialog with a couple of entry widgets which would allow the user to set certain metadata from a predefined set.

```python
from dataclasses import dataclass

@dataclass
class DiagramMetadata:
    """Metadata for a Gaphor project file."""
    created_at: str
    created_by: str
    description: str
    extensions: Dict[str, str]
    license: str
    revision: str
    tags: Set[str]
    updated_at: str
    website: str
```

```xml
<?xml version="1.0" encoding="utf-8"?>
<gaphor xmlns="http://gaphor.sourceforge.net/model" version="3.0" gaphor-version="2.9.2">
  <!-- ... -->
  <metadata>
    <createdAt>2021-06-11</createdAt>
    <createdBy>Alice A. Gaphor</createdBy>
    <description>Prototype model SX-1 electric vehicle.</description>
    <extensions>
      <extension key="author_fingerprint">EB85 BB5F A33A 75E1 5E94 4E63 F231 550C 4F47 E38E</extension>
      <extension key="hmac">de7c9b85b8b78aa6bc8a7a36f70a90701c9db4d9</extension>
      <extension key="git_commit">ed054a38</extension>
    </extensions>
    <license>CC-PDDC</license>
    <revision>0.1A</revision>
    <tags>ev;vehicle;prototype</tags>
    <updatedAt>2021-12-06</updatedAt>
    <website>https://gaphor.org/</website>
  </metadata>
  <!-- ... -->
</gaphor>
```

**Comment by amolenaar** [comment]
I would like to treat meta data as a normal model element. You should be able to add it to a diagram, edit it in the editor pane on the right and be able to drag it in from the tree view to any model. Some fields are model-wide (e.g. license) and some are diagram specific (e.g. diagram description).

**Comment by vanillajonathan** [comment]
But you might also want invisible metadata that does not display in the diagram but is available to search engines, indexers, and tools, such as tags or arbitrary metadata for things such as author GPG fingerprint, HMAC signature, Git commit id, etc.

**Comment by amolenaar** [comment]
Some inspiration on metadata. I suppose Dublin Core is the standard here.

* https://www.dublincore.org/specifications/dublin-core/ ([xml schemas](https://www.dublincore.org/schemas/xmls/))
* https://books.google.nl/books?id=nNQqDgAAQBAJ&pg=PA25&lpg=PA25&dq=xml+metadata+author+version&source=bl&ots=SfagvyV60X&sig=ACfU3U0kJzkXgGyIDD-uZhqpvkfaCSe8wg&hl=en&sa=X&ved=2ahUKEwi8yreRl574AhWCKewKHXBJCvgQ6AF6BAgoEAM#v=onepage&q=xml%20metadata%20author%20version&f=false

**Comment by amolenaar** [comment]
A first attempt:

![image](https://user-images.githubusercontent.com/96249/218269642-51f37fc9-5836-4996-b472-bfc42c9fc480.png)

**Comment by vanillajonathan** [comment]
Good work!

I am also interested in having invisible metadata that is not visible on the diagram for arbitrary data such as tags, author GPG fingerprint, HMAC signature, etc.

**Comment by danyeaw** [comment]
Fixed in #2082.
